src/platforms/common/usb_serial.c

```diff
@@ -77,7 +77,7 @@ static void debug_serial_receive_callback(usbd_device *dev, uint8_t ep);
 static bool debug_serial_send_complete = true;
 #endif
 
-#if defined(ENABLE_DEBUG) && defined(PLATFORM_HAS_DEBUG)
+#if (ENABLE_DEBUG == 1) && defined(PLATFORM_HAS_DEBUG)
```

Comment:
I prefer parantheses for explicit operator precedence (as suggested by MISRA), but I will change it.

Reply:
Ok, we're going to call MISRA out as doing a bad here: The compiler, outside of preprocessor macros, has a warning for accidental assignment, which the parens entirely disable. This has lead to several bugs in BMD that we had to fix that we only found by *removing* the parens to make the compiler chirp - ie, the extra parens are bugprone, turning off the only help you can get from the compiler.

It's a little less bad in the preprocessor situation in that it just muddies the expression needlessly (the preprocessor can't assign in this context), but it's inconsistent with the rest of the code base and comes from a place of worrying about && and || precedence vs equality (equality always binds tighter, making it a non-problem) while disregarding the side effects of the policy (inhibited -Wparentheses).

This is perhaps something to keep in mind as putting in the parens around a complete expression - ie `(a == b && b == c)` is fine and doesn't inhibit this warning - it's specifically around the comparisons that it's a problem.

Reply:
I see, interesting

Reply:
We have to admit that we've seen that and whenever it's been done we find it leads to very difficult to read code as all the conditionals are backwards, making one have to do serious mental gymnastics, particularly for inequalities - we honestly favour writing things the normal way around and putting `-Wparentheses` on, banning all assignment in comparisons in the process by virtue of the compiler enforcing it.

We feel it leads to more natural code (a better expression of intent) while keeping the likelihood of introducing the class of bugs MISRA is trying to fix there, as minimal as MISRA might - hence our favouring this other approach. (You might have noticed, we strongly favour having the tools help us rather than trying to write code as if the tools can't without help and it's beholden on the programmer instead, because the programmer can and will make mistakes)

Reply:
Yes, I agree that it looks quite strange with the conditionals in that order. Using `-Wparantheses` is certainly a better solution to the problem.

Regarding the use of tools and avoiding programmer errors, we enforce MISRA rules by means of something like `cppcheck`, but it is better if the compiler can do it by itself.